### PR TITLE
LPS-131159 Implement Image Editor in item selector

### PR DESCRIPTION
**Pull request by carloslancha**
https://issues.liferay.com/browse/LPS-131159

This is the implementation of the new Image Editor in item selector. It's not done in the best way since item selector implementation it's pretty overcomplicated and it's not prepared to be extended with this kind of stuff. IMHO it should be re-thinked and re-implemented so it fits with our needs in the most easy and with the most sense possible.

In the meantime this is the most non-invasive implementation I could find.

To test it:

- Go anywhere with an item selector (i.e. Blogs)
- Go to an image preview (by clicking on the "eye" icon on the item) or upload a new one
- Click on the "pencil" icon in the top bar of the image preview
- Edit. Save. Profit.

### CODE REVIEW

**Comment by AliciaGarciaGarcia** [comment]
Just started reviewing :)

**Comment by carloslancha** [comment]
Rebased with master and SF. It should pass now

**Comment by carloslancha** [comment]
ci:test:sf

**Comment by AliciaGarciaGarcia** [comment]
Just started reviewing :)

**Comment by carloslancha** [comment]
ci:test:relevant

**Comment by AliciaGarciaGarcia** [comment]
Hi @carloslancha 
I found two issues that I do not know if are related with this story:

One of them is with the page attachments on a wiki. If you try to send the file, there ir a issue that do not attach the file, but it create the file on D&M

https://user-images.githubusercontent.com/47524751/117644678-06a4d100-b18a-11eb-988a-17dee53eae36.mp4

On the Knowledge base:
There is an error on the console when you edit an image but the image and the image edited are saved.

`GET http://localhost:8080/group/guest/~/control_panel/[object%20Object] 400`

https://user-images.githubusercontent.com/47524751/117644977-57b4c500-b18a-11eb-81c3-fd0ba7343d6e.mp4

Thanks!

**Comment by carloslancha** [comment]
@AliciaGarciaGarcia about the second issue it seems to be related with the response on the edit image call being different than the received on upload, and it looks like the integration there with ckeditor is prepared to deal with the first one.

Upload response contains `resolvedValue` with a url as value, but image edition response return no `resolvedValue`, so it's created in the client as a stringified JSON. Could you please take a look to compare both responses so they work the same way?

I'll take a look to the other issue in the meantime.

Thx!

**Comment by carloslancha** [comment]
@AliciaGarciaGarcia I may no explained it correctly:

### Wiki

Upload image response:
```
{"file":{"attributeDataImageId":"data-image-id","randomId":"","groupId":"20126","resolvedValue":"{\"groupId\":\"20126\",\"title\":\"cat2 (2).jpg\",\"type\":\"document\",\"uuid\":\"7ed39603-1949-d431-bdbf-c54fd887d29a\",\"fileEntryId\":\"40314\",\"url\":\"/documents/portlet_file_entry/20126/cat2+%282%29.jpg/7ed39603-1949-d431-bdbf-c54fd887d29a?imagePreview=1\"}","mimeType":"image/jpeg","title":"cat2 (2).jpg","type":"document","uuid":"7ed39603-1949-d431-bdbf-c54fd887d29a","fileEntryId":"40314","url":"http://localhost:8080/documents/portlet_file_entry/20126/cat2+%282%29.jpg/7ed39603-1949-d431-bdbf-c54fd887d29a"},"success":true}
```

Save edited image response
```
{"file":{"attributeDataImageId":"data-image-id","randomId":"","groupId":"20126","resolvedValue":"{\"groupId\":\"20126\",\"title\":\"cat2 (2) (1).jpg\",\"type\":\"document\",\"uuid\":\"422377dd-ac12-15e7-7363-3e633bbb29d4\",\"fileEntryId\":\"40322\",\"url\":\"/documents/portlet_file_entry/20126/cat2+%282%29+%281%29.jpg/422377dd-ac12-15e7-7363-3e633bbb29d4?imagePreview=1\"}","mimeType":"image/png","title":"cat2 (2) (1).jpg","type":"document","uuid":"422377dd-ac12-15e7-7363-3e633bbb29d4","fileEntryId":"40322","url":"http://localhost:8080/documents/portlet_file_entry/20126/cat2+%282%29+%281%29.jpg/422377dd-ac12-15e7-7363-3e633bbb29d4"},"success":true}
```

### KB

Upload image response:
```
{"file":{"attributeDataImageId":"data-image-id","randomId":"","groupId":"20126","resolvedValue":"/documents/20126/0/arctic+fox+%281%29.jpg/4495cb21-8847-c077-e4e4-1d7d61d84f3b?t=1620653418931","mimeType":"image/jpeg","title":"arctic fox (1).jpg","type":"document","uuid":"4495cb21-8847-c077-e4e4-1d7d61d84f3b","fileEntryId":"40330","url":"http://localhost:8080/documents/20126/0/arctic+fox+%281%29.jpg/4495cb21-8847-c077-e4e4-1d7d61d84f3b?version=1.0&t=1620653418931"},"success":true}
```

Save edited image response:
```
{"file":{"attributeDataImageId":"data-image-id","randomId":"","groupId":"20126","mimeType":"image/png","title":"Dev21-Computer large (3).jpg","type":"document","uuid":"67a37cbe-78dd-445a-dd7c-9e69bd1f69c5","fileEntryId":"40341","url":"http://localhost:8080/documents/20126/0/Dev21-Computer+large+%283%29.jpg/67a37cbe-78dd-445a-dd7c-9e69bd1f69c5?version=1.0&t=1620653702376"},"success":true}
```

As you can see in Wiki we receive, in both cases, a similar `resolveValue` (a stringified JSON in this case), while in KB we receive from the upload image response a relative url as `resolvedValue` but nothing for dat property from the edited image response.

In that case (`resolvedValue` is undefined) the previously implemented logic is to create in client the stringified JSON value from the rest of the properties, but in this case what ckeditor expects is that relative url, so we need to return the same stuff in both calls.

File: modules/apps/item-selector/item-selector-taglib/src/main/resources/META-INF/resources/repository_entry_browser/js/ItemSelectorRepositoryEntryBrowser.es.js [hunk 1]
```diff
@@ -157,7 +158,7 @@ class ItemSelectorRepositoryEntryBrowser extends PortletBase {
 
 					Liferay.componentReady('ItemSelectorPreview').then(() => {
 						Liferay.fire('updateCurrentItem', {
-							url: itemFileUrl,
+							...itemFile,
```
[the remark this answers is not in the document]
Comment (carloslancha) [reply modules/apps/item-selector/item-selector-taglib/src/main/resources/META-INF/resources/repository_entry_browser/js/ItemSelectorRepositoryEntryBrowser.es.js:161]: Yes, we need `resolvedValue` as well as other properties. It was failing before in some cases. I just want to stress what I said in the description:
> IMHO it should be re-thinked and re-implemented so it fits with our needs in the most easy and with the most sense possible.

**Comment by carloslancha** [comment]
> I think that in one point or another we are do not sending the "returnType" parameter that on the upload we must be sending. But I do not have more information on this point.

In the response of the image edition we're not returning any `returnType`, but since we're creating a new item with the previous values (the ones returned by upload) mixed with edited returned ones we have it defined in client.

**Comment by boton** [comment]
FYI

Some `item-selector/item-selector-taglib` are failing.

image

**Comment by carloslancha** [comment]
@AliciaGarciaGarcia  Sorry, I didn't understand correctly, I thought you were saying that the service was not sending the `returnType`. Let's stop saying "we're" and be more clear about everything since this is getting pretty confusing.

So, you need the client to send to the save edited image service the same returnType that's being passed to the upload service?

**Comment by AliciaGarciaGarcia** [comment]
Yes @carloslancha , I need the `returnType` param to generate the `resolvedValue`

**Comment by AliciaGarciaGarcia** [comment]
Just started reviewing :)

**Comment by boton** [comment]
Work for me in Web Content but doesn't work for me in blogs, I try with clean ant all because I'm not sure if this a local issue and doesn't work on blogs.


https://user-images.githubusercontent.com/67901/117859085-dbf07080-b28e-11eb-8c5a-74e603b51c09.mp4

**Comment by carloslancha** [comment]
> Work for me in Web Content but doesn't work for me in blogs
It's working for me.

> The save button isn't disabled while is generating the image and you can click on save button a lot of time.
???

**Comment by boton** [comment]
> The save button isn't disabled while is generating the image and you can click on save button a lot of time.

> ???

The `save` button still enable when we are fetching the image and we can click on it several times, maybe its better to disable it when we are fetching avoiding parallel request

Parallel request of image editor:
image

**Comment by carloslancha** [comment]
Tests failing since jest is throwing an error when ImageEditor tries to import a css file. Addressed in https://github.com/liferay/liferay-frontend-projects/issues/534

**Comment by boton** [comment]
I'm tested with clear DB and ant all with only this branch and doesn´t work for me in Blogs, maybe it's a local issue.

The failing tests are fixed only this minus issue https://github.com/liferay-lima/liferay-portal/pull/1802#issuecomment-839841999 caught by @carloslancha and it is working on progress.

LGTM

**Comment by carloslancha** [comment]
@boton could you please add the steps to try the reproduce the issue you're having in blogs?

**Comment by boton** [comment]
@carloslancha sure

1. Go to Content & Data > blogs > New entry [ + ]
2. Fill title field
3. Try to add a new image or edit an existing one doesn’t work for me

You can see in https://github.com/liferay-lima/liferay-portal/pull/1802#issuecomment-838871967 video

**Comment by carloslancha** [comment]
J30x3BITMn

@boton ¯\_(ツ)_/¯

**Comment by boton** [comment]
> J30x3BITMn
> 
> @boton ¯_(ツ)_/¯

Looks beautiful, nothing to add. I will try an ant all with deep clean at the end of the day.

Ready to go!

**Review by boton — APPROVED** [review]

**Comment by carloslancha** [comment]
@izaera fixed the problem in https://github.com/liferay/liferay-frontend-projects/pull/537 but we need that fix to be merged and released so I'm ignoring the failing tests so we can move forward.

@adolfopa please take a look to Java code and, if tests pass, forward the pr. Thx guys!

**Comment by carloslancha** [comment]
ci:test:sf

**Comment by carloslancha** [comment]
ci:test:relevant

**Review by adolfopa — APPROVED** [review]

**Comment by adolfopa** [comment]
ci:forward

**Pull request closed without merging**